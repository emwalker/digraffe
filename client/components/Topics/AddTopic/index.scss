.container {
  input, button {
    display: inline-block;
  }

  .topic-name {
    width: 75%;
    margin-right: 1em;
  }
}
