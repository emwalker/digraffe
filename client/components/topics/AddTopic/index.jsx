// @flow
import React, { Component } from 'react'
import { Button, FormGroup, Input } from 'reactstrap'

import createTopicMutation from '../../../mutations/createTopicMutation'
import './index.scss'

type Props = {
  organization: {
    id: string,
    resourceId: string,
  },
  relay: {
    environment: Object,
  },
}

type State = {
  name: string,
}

class AddTopic extends Component<Props, State> {
  state = {
    name: '',
  }

  onSubmit = () => {
    const {
      id: orgId,
      resourceId: organizationResourceId,
    } = this.props.organization
    createTopicMutation(
      this.props.relay.environment,
      orgId,
      {
        organizationResourceId,
        name: this.state.name,
      },
    )
    this.setState({ name: '' })
  }

  updateName = (event: Object) => {
    this.setState({ name: event.currentTarget.value })
  }

  render = () => (
    <div stylename="container">
      <FormGroup>
        <Input
          className="topic-name test-topic-name"
          placeholder="Add a topic"
          value={this.state.name}
          onChange={this.updateName}
        />
        <Button size="sm" onClick={this.onSubmit}>Submit</Button>
      </FormGroup>
    </div>
  )
}

export default AddTopic
