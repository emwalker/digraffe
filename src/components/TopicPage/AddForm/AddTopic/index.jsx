// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import type { Relay, RepositoryType, TopicType, UserType } from 'components/types'
import upsertTopicMutation from 'mutations/upsertTopicMutation'

/* eslint jsx-a11y/label-has-for: 0 */

type Props = {
  disabled?: boolean,
  relay: Relay,
  topic: TopicType,
  viewer: UserType,
}

type State = {
  name: string,
}

class AddTopic extends Component<Props, State> {
  static defaultProps = {
    disabled: true,
  }

  state = {
    name: '',
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.createTopic()
  }

  get selectedRepo(): ?RepositoryType {
    return this.props.viewer.selectedRepository
  }

  get orgLogin(): ?string {
    const repo = this.selectedRepo
    if (!repo)
      return null
    return repo.organization.login
  }

  get relayConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_childTopics',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'topicEdge',
    }]
  }

  updateName = (event: Object) => {
    this.setState({ name: event.currentTarget.value })
  }

  createTopic() {
    const repo = this.selectedRepo
    const repoName = repo ? repo.name : null

    upsertTopicMutation(
      this.props.relay.environment,
      this.relayConfigs,
      {
        name: this.state.name,
        repositoryName: repoName,
        organizationLogin: this.orgLogin,
        topicIds: [this.props.topic.id],
      },
    )
    this.setState({ name: '' })
  }

  render = () => (
    <dl className="form-group">
      <dt>
        <label htmlFor="create-topic-name">Add subtopic</label>
      </dt>
      <dd>
        <input
          className="form-control test-topic-name input-sm"
          disabled={this.props.disabled}
          id="create-topic-name"
          onChange={this.updateName}
          onKeyPress={this.onKeyPress}
          placeholder="Name or description"
          value={this.state.name}
        />
      </dd>
    </dl>
  )
}

export default createFragmentContainer(AddTopic, graphql`
  fragment AddTopic_viewer on User {
    selectedRepository {
      name
      organization {
        login
      }
    }
  }

  fragment AddTopic_topic on Topic {
    id
  }
`)
