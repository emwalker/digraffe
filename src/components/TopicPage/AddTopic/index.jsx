// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import upsertTopicMutation from 'mutations/upsertTopicMutation'
import type { RelayProps } from '../../types'

/* eslint jsx-a11y/label-has-for: 0 */

type State = {
  name: string,
}

type Props = RelayProps & {
  viewer: {
    defaultRepository: {
      id: ID,
    },
  },
}

class AddTopic extends Component<Props, State> {
  state = {
    name: '',
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.createTopic()
  }

  get relayConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_childTopics',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'topicEdge',
    }]
  }

  get repositoryId(): ?string {
    if (this.props.viewer)
      return this.props.viewer.defaultRepository.id
    return null
  }

  updateName = (event: Object) => {
    this.setState({ name: event.currentTarget.value })
  }

  createTopic() {
    upsertTopicMutation(
      this.props.relay.environment,
      this.relayConfigs,
      {
        name: this.state.name,
        repositoryId: this.repositoryId,
        topicIds: [this.props.topic.id],
      },
    )
    this.setState({ name: '' })
  }

  render = () => (
    <div>
      <dl className="form-group">
        <dt>
          <label htmlFor="create-topic-name">Add subtopic</label>
        </dt>
        <dd>
          <input
            className="form-control test-topic-name input-sm"
            id="create-topic-name"
            onChange={this.updateName}
            onKeyPress={this.onKeyPress}
            placeholder="Name or description"
            value={this.state.name}
          />
        </dd>
      </dl>
    </div>
  )
}

export default createFragmentContainer(AddTopic, graphql`
  fragment AddTopic_topic on Topic {
    id
  }
`)
