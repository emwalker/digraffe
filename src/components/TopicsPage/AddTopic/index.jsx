// @flow
import React, { Component } from 'react'

import createTopicMutation from '../../../mutations/createTopicMutation'

type Props = {
  organization: {
    id: string,
    resourceId: string,
  },
  relay: {
    environment: Object,
  },
}

type State = {
  name: string,
}

class AddTopic extends Component<Props, State> {
  state = {
    name: '',
  }

  onSubmit = () => {
    const {
      id: orgId,
      resourceId: organizationId,
    } = this.props.organization
    createTopicMutation(
      this.props.relay.environment,
      orgId,
      {
        organizationId,
        name: this.state.name,
      },
    )
    this.setState({ name: '' })
  }

  updateName = (event: Object) => {
    this.setState({ name: event.currentTarget.value })
  }

  render = () => (
    <div>
      <dl className="form-group">
        <dd>
          <input
            className="form-control test-topic-name"
            placeholder="Topic"
            value={this.state.name}
            onChange={this.updateName}
          />
        </dd>
        <button
          size="sm"
          className="d-inline-block btn btn-secondary mt-2"
          onClick={this.onSubmit}
        >
          Add
        </button>
      </dl>
    </div>
  )
}

export default AddTopic
