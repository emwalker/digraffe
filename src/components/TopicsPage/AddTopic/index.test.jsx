import React from 'react'
import { shallow } from 'enzyme'
import AddTopic from './index'

describe('<AddTopic />', () => {
  const organization = {
    resourceId: 'organization:tyrell',
  }

  const wrapper = shallow(
    <AddTopic
      organization={organization}
    />,
  )

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('includes an input for the name', () => {
    expect(wrapper.find('input.test-topic-name')).toHaveLength(1)
  })
})
