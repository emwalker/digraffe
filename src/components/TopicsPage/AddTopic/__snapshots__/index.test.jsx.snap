// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTopic /> renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddTopic
    organization={
      Object {
        "resourceId": "organization:tyrell",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <dl
        className="form-group"
      >
        <dd>
          <input
            className="form-control test-topic-name"
            onChange={[Function]}
            placeholder="Topic"
            value=""
          />
        </dd>
        <button
          className="d-inline-block btn btn-secondary mt-2"
          onClick={[Function]}
          size="sm"
        >
          Add
        </button>
      </dl>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <dd>
            <input
              className="form-control test-topic-name"
              onChange={[Function]}
              placeholder="Topic"
              value=""
            />
          </dd>,
          <button
            className="d-inline-block btn btn-secondary mt-2"
            onClick={[Function]}
            size="sm"
          >
            Add
          </button>,
        ],
        "className": "form-group",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              className="form-control test-topic-name"
              onChange={[Function]}
              placeholder="Topic"
              value=""
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control test-topic-name",
              "onChange": [Function],
              "placeholder": "Topic",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "dd",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add",
            "className": "d-inline-block btn btn-secondary mt-2",
            "onClick": [Function],
            "size": "sm",
          },
          "ref": null,
          "rendered": "Add",
          "type": "button",
        },
      ],
      "type": "dl",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <dl
          className="form-group"
        >
          <dd>
            <input
              className="form-control test-topic-name"
              onChange={[Function]}
              placeholder="Topic"
              value=""
            />
          </dd>
          <button
            className="d-inline-block btn btn-secondary mt-2"
            onClick={[Function]}
            size="sm"
          >
            Add
          </button>
        </dl>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <dd>
              <input
                className="form-control test-topic-name"
                onChange={[Function]}
                placeholder="Topic"
                value=""
              />
            </dd>,
            <button
              className="d-inline-block btn btn-secondary mt-2"
              onClick={[Function]}
              size="sm"
            >
              Add
            </button>,
          ],
          "className": "form-group",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="form-control test-topic-name"
                onChange={[Function]}
                placeholder="Topic"
                value=""
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control test-topic-name",
                "onChange": [Function],
                "placeholder": "Topic",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "dd",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add",
              "className": "d-inline-block btn btn-secondary mt-2",
              "onClick": [Function],
              "size": "sm",
            },
            "ref": null,
            "rendered": "Add",
            "type": "button",
          },
        ],
        "type": "dl",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
